### ISSUE TITLE
go/parser: parser creates Stmt that causes token.File.Offset(stmt.End) to panic

### ISSUE THREAD

**Issue by adonovan** [golang/go#57490]
Given an input of `src`, the parser reserves a block of the token.Pos space of exactly len(src), even though it may synthesize virtual tokens (e.g. close brackets) at the EOF position during error recovery. The End position of the incomplete syntax node is then computed as Rparen + len(")"), which is out of bounds. See https://github.com/golang/go/issues/57484 for a minimal example.

The parser should reserve one extra byte in its call to FileSet.AddFile. Or use [safePos](https://cs.opensource.google/go/go/+/master:src/go/parser/parser.go;l=428-436) in many more places.

@mdempsky @findleyr

**Comment by griesemer** [comment]
Panicking example: https://go.dev/play/p/JXREnQBsSPe

**Comment by griesemer** [comment]
@adonovan I looked into this:
- As far as I can tell, where the parser reports end positions needed for error messages, it already uses safePos.
- There's still the issue of the AST storing positions past the end, caused by syntax errors.

I see several approaches here:
1. We remove the various assertions (panics) in go/token/position.go in favor of an automatic bracketing of the positions: if we are "before" the file, we set the position (or offset) to the start of the file; if we are "after" the file, we set the position (offset) to the end of the file.
2. We add a few bytes to the end of a file as a safety measure, and we adjust the tests that check for exact sizes.
3. We don't do anything because tools have already been adjusted.

Approach 2) seems problematic because it changes a major invariant with unforeseeable consequences. 3) may be unsatisfactory because we don't know if we fixed all places already.

I think 1) may be doable: removing the panics should not break existing code (parser.safePos notwithstanding), and if we have out-of-bounds positions, they will be automatically corrected.

If we decide to do 1) we should do in early in cycle to catch problems ASAP.
